observer/baseNodeProvider.go

```diff
@@ -25,8 +26,37 @@ func (bop *baseNodeProvider) initNodesMaps(nodes []*data.NodeData) error {
 
 	bop.mutNodes.Lock()
 	bop.nodes = newNodes
-	bop.allNodes = nodes
+	bop.allNodes = initAllNodesSlice(newNodes)
 	bop.mutNodes.Unlock()
 
 	return nil
 }
+
+func initAllNodesSlice(nodesOnShards map[uint32][]*data.NodeData) []*data.NodeData {
+	sliceToReturn := make([]*data.NodeData, 0)
+	shardIDs := make([]uint32, 0)
+	for shardID := range nodesOnShards {
+		shardIDs = append(shardIDs, shardID)
+	}
+	sort.SliceStable(shardIDs, func(i, j int) bool {
+		return shardIDs[i] < shardIDs[j]
+	})
+
+	finishedShards := make(map[uint32]struct{})
+	for i := 0; ; i++ {
+		for _, shardID := range shardIDs {
+			if i >= len(nodesOnShards[shardID]) {
+				finishedShards[shardID] = struct{}{}
```

Comment:
This solution will have the same behavior as the old code in the following scenario:
shard 0: 1 observer (that might temporarily fail) - this observer is A
shard 1: 1 observer (doesn't have have the data) - this observer is B
shard 2: 20 observers (do not have the data) - these observers are C1...C20

Request sent can only be resolved by the shard 0 observer 
I think it would have been better that the "finished shards" will reiterate their lists continuously.
From the above example, the list should be:
A, B, C1, A, B, C2, A, B, C3 ...... A, B, C20
What do you think? @bogdan-rosianu @AdoAdoAdo

Reply:
refactored as discussed